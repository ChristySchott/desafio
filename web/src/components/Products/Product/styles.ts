import styled from 'styled-components';

export const Container = styled.li`
  margin-top: 25px;
  width: 135px;

  @media (min-width: 768px) {
    width: 150px;
  }

  @media (min-width: 1024px) {
    width: 170px;
  }
`;

export const Image = styled.div`
  border: 1px solid #ddd;

  img {
    width: 100%;
  }
`;

export const Infos = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  align-items: center;
  text-align: center;

  height: 100px;
  padding: 10px 0;

  span {
    text-transform: uppercase;
    font-size: 1.6rem;
  }

  span:nth-child(2) {
    text-decoration: line-through;
    color: #999;
    font-size: 1.5rem;
  }

  strong {
    color: #c61a28;
    font-weight: 600;
  }
`;
