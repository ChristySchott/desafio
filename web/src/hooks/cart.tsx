import React, {
  createContext,
  useState,
  useCallback,
  useContext,
  useEffect,
} from 'react';

export interface ProductInterface {
  id: number;
  sku: string;
  path: string;
  name: string;
  image: string;
  price: number;
  specialPrice?: number;
  quantity: number;
}

interface CartContext {
  products: ProductInterface[];
  addToCart(item: Omit<ProductInterface, 'quantity'>): void;
  increment(id: number): void;
  decrement(id: number): void;
}

const CartContext = createContext<CartContext | null>(null);

const CartProvider: React.FC = ({ children }) => {
  const [products, setProducts] = useState<ProductInterface[]>([]);

  useEffect(() => {
    async function loadProducts(): Promise<void> {
      const storagedProducts = localStorage.getItem('@Webjump:products');

      if (storagedProducts) {
        setProducts([...JSON.parse(storagedProducts)]);
      }
    }

    loadProducts();
  }, []);

  const addToCart = useCallback(
    async product => {
      const productExists = products.find(item => item.id === product.id);

      if (productExists) {
        setProducts(
          products.map(p =>
            p.id === product.id ? { ...product, quantity: p.quantity + 1 } : p,
          ),
        );
      } else {
        setProducts([...products, { ...product, quantity: 1 }]);
      }

      localStorage.setItem('@Webjump:products', JSON.stringify(products));
    },
    [products],
  );

  const increment = useCallback(
    async id => {
      const updatedProducts = products.map(product =>
        product.id === id
          ? { ...product, quantity: product.quantity + 1 }
          : product,
      );

      setProducts(updatedProducts);

      localStorage.setItem(
        '@Webjump:products',
        JSON.stringify(updatedProducts),
      );
    },
    [products],
  );

  const decrement = useCallback(
    async id => {
      const updatedProducts = products.map(product =>
        product.id === id
          ? { ...product, quantity: product.quantity - 1 }
          : product,
      );

      setProducts(updatedProducts);

      localStorage.setItem(
        '@Webjump:products',
        JSON.stringify(updatedProducts),
      );
    },
    [products],
  );

  const value = React.useMemo(
    () => ({ addToCart, increment, decrement, products }),
    [products, addToCart, increment, decrement],
  );

  return <CartContext.Provider value={value}>{children}</CartContext.Provider>;
};

function useCart(): CartContext {
  const context = useContext(CartContext);

  if (!context) {
    throw new Error(`useCart must be used within a CartProvider`);
  }

  return context;
}

export { CartProvider, useCart };
