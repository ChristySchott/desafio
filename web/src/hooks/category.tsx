import React, { createContext, useContext, useState } from 'react';

interface CategoryContextData {
  category: string;
  setCategory: React.Dispatch<React.SetStateAction<string>>;
}

const CategoryContext = createContext<CategoryContextData>(
  {} as CategoryContextData,
);

const CategoryProvider: React.FC = ({ children }) => {
  const [category, setCategory] = useState('calcados');

  return (
    <CategoryContext.Provider
      value={{
        category,
        setCategory,
      }}
    >
      {children}
    </CategoryContext.Provider>
  );
};

function useCategory(): CategoryContextData {
  const context = useContext(CategoryContext);

  if (!context) {
    throw new Error('useCategory must be used within an CategoryContext');
  }

  return context;
}

export { CategoryProvider, useCategory };
