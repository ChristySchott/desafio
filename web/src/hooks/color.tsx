import React, { createContext, useContext, useState } from 'react';

interface ColorContextData {
  colorToFilter: string;
  setColorToFilter: React.Dispatch<React.SetStateAction<string>>;
}

const ColorContext = createContext<ColorContextData>({} as ColorContextData);

const ColorProvider: React.FC = ({ children }) => {
  const [colorToFilter, setColorToFilter] = useState('');

  return (
    <ColorContext.Provider
      value={{
        colorToFilter,
        setColorToFilter,
      }}
    >
      {children}
    </ColorContext.Provider>
  );
};

function useColorFilter(): ColorContextData {
  const context = useContext(ColorContext);

  if (!context) {
    throw new Error('useColorFilter must be used within an ColorContext');
  }

  return context;
}

export { ColorProvider, useColorFilter };
