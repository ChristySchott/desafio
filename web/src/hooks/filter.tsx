import React, { createContext, useContext, useState } from 'react';

interface FilterContextData {
  filter: string;
  setFilter: React.Dispatch<React.SetStateAction<string>>;
}

const FilterContext = createContext<FilterContextData>({} as FilterContextData);

const FilterProvider: React.FC = ({ children }) => {
  const [filter, setFilter] = useState('');

  return (
    <FilterContext.Provider
      value={{
        filter,
        setFilter,
      }}
    >
      {children}
    </FilterContext.Provider>
  );
};

function useFilter(): FilterContextData {
  const context = useContext(FilterContext);

  if (!context) {
    throw new Error('useFilter must be used within an FilterContext');
  }

  return context;
}

export { FilterProvider, useFilter };
