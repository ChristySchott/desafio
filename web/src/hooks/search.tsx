import React, { createContext, useContext, useState } from 'react';

interface SearchContextData {
  search: string;
  setSearch: React.Dispatch<React.SetStateAction<string>>;
}

const SearchContext = createContext<SearchContextData>({} as SearchContextData);

const SearchProvider: React.FC = ({ children }) => {
  const [search, setSearch] = useState('');

  return (
    <SearchContext.Provider
      value={{
        search,
        setSearch,
      }}
    >
      {children}
    </SearchContext.Provider>
  );
};

function useSearch(): SearchContextData {
  const context = useContext(SearchContext);

  if (!context) {
    throw new Error('useSearch must be used within an SearchContext');
  }

  return context;
}

export { SearchProvider, useSearch };
