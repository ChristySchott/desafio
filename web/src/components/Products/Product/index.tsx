/* eslint-disable import/no-dynamic-require */
/* eslint-disable global-require */
import React from 'react';

import formatValue from 'utils/formatValue';
import { ProductInterface, useCart } from 'hooks/cart';
import { Container, Image, Infos } from './styles';

interface ProductProps {
  imageUrl: string;
  alt: string;
  name: string;
  offer?: number;
  price: number;
  item: ProductInterface;
}

const Product: React.FC<ProductProps> = ({ alt, name, offer, price, item }) => {
  const { addToCart } = useCart();

  function handleAddToCart(product: ProductInterface): void {
    addToCart(product);
  }

  return (
    <Container>
      <a href="/">
        <Image>
          <img src={require('assets/products/shoes-6.jpg')} alt={alt} />
        </Image>
      </a>

      <Infos>
        <span>{name}</span>
        {!!offer && <span>{formatValue(offer)}</span>}
        <strong>{formatValue(price)}</strong>
      </Infos>

      <button type="button" onClick={() => handleAddToCart(item)}>
        Comprar
      </button>
    </Container>
  );
};

export default Product;
