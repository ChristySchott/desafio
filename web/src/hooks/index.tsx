import React from 'react';

import { FilterProvider } from './filter';
import { ColorProvider } from './color';
import { CartProvider } from './cart';

const AppProvider: React.FC = ({ children }) => {
  return (
    <FilterProvider>
      <ColorProvider>
        <CartProvider>{children}</CartProvider>
      </ColorProvider>
    </FilterProvider>
  );
};

export default AppProvider;
