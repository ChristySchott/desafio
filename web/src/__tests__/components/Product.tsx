import React from 'react';
import { render } from '@testing-library/react';

import { CartProvider } from 'hooks/cart';
import Product from '../../components/Products/Product';

describe('<Product />', () => {
  it('should render with no error', () => {
    const product = {
      id: 1,
      sku: 'value',
      path: 'value',
      name: 'value',
      image: 'value',
      price: 1,
      specialPrice: 1,
      quantity: 1,
    };

    const { container } = render(
      <CartProvider>
        <Product imageUrl="" alt="" name="" price={2} item={product} />
      </CartProvider>,
    );

    expect(container).toBeTruthy();

    expect(container.firstChild).toMatchSnapshot();
  });
});
