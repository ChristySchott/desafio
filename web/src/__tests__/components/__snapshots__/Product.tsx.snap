// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Product /> should render with no error 1`] = `
<li
  class="sc-bdnylx dcpTIQ"
>
  <a
    href="/"
  >
    <div
      class="sc-gtssRu hQwYBx"
    >
      <img
        alt=""
        src="shoes-6.jpg"
      />
    </div>
  </a>
  <div
    class="sc-dlnjPT kMBBxv"
  >
    <span />
    <strong>
      R$2.00
    </strong>
  </div>
  <button
    type="button"
  >
    Comprar
  </button>
</li>
`;
